osbuild/util/fscache.py

```diff
@@ -0,0 +1,1008 @@
+"""File System Cache
+
+This module implements a data cache that uses the file system to store data
+as well as protect parallel access. It implements automatic cache management
+and allows purging the cache during runtime, pruning old entries and keeping
+the cache under a given limit.
+"""
+
+# pylint: disable=too-many-lines
+
+import contextlib
+import ctypes
+import errno
+import json
+import os
+import uuid
+from typing import Any, Dict, Literal, NamedTuple, Optional, Tuple, Union
+
+from osbuild.util import ctx, linux, rmrf
+
+__all__ = [
+    "FsCache",
+    "FsCacheInfo",
+]
+
+
+class FsCacheInfo(NamedTuple):
+    """File System Cache Information
+
+    This type represents static cache information. It is an immutable named
+    tuple and used to query or set the configuration of a cache.
+
+    creation_boot_id - Hashed linux boot-id at the time of cache-creation
+    maximum_size - Maximum cache size in bytes, or "unlimited"
+    """
+
+    creation_boot_id: Optional[str] = None
+    maximum_size: Optional[Union[int, Literal["unlimited"]]] = None
+
+    @classmethod
+    def from_json(cls, data: Any) -> "FsCacheInfo":
+        """Create tuple from parsed JSON
+
+        This takes a parsed JSON value and converts it into a tuple with the
+        same information. Unknown fields in the input are ignored. The input
+        is usually taken from `json.load()` and similar.
+        """
+
+        if not isinstance(data, dict):
+            return cls()
+
+        creation_boot_id = None
+        maximum_size: Optional[Union[int, Literal["unlimited"]]] = None
```

Comment:
I think it would be nice to have `Union[int, Literal["unlimited"]]` defined as a top level type so it can be re-used, i.e. in `objectstore.ObjectStore.maximum_size`.

Reply:
Also bad news:
`ImportError: cannot import name 'Literal'`

```
Literal[...] cannot be subclassed. At runtime, an arbitrary value is allowed as type argument to Literal[...], but type checkers may impose restrictions. See PEP 586 for more details about literal types.

New in version 3.8.
```
**New in version 3.8.**

Reply:
I wonder if we can away with a tristate, in the API. For the file it would be nice to keep `unlimited` but `Optional[Union[int, str]]`: is a bit annoying to work with

- `None`: size not set, unlimited
- `0`: save nothing
- `int`: size limit

Reply:
Yeah, that sounds ok.

Reply:
Ok, going with the Union over `int` and `str`, since python does not allow nesting `Optional` types, and hence the upper layer breaks because it adds an `Optional` layer to the stored type to decide whether it is present in the InfoClass.

It is actually ok from the code-perspective, almost no difference to the `Literal` handling.

Reply:
Yeah that is what I did for my quick-fix, except now you have to handle an additional case:
```patch
@@ -592,6 +592,8 @@ class FsCache(contextlib.AbstractContextManager, os.PathLike):
         # Parse `maximum-size` into internal representation.
         if info.maximum_size == "unlimited":
             self._info_maximum_size = -1
+        elif isinstance(info.maximum_size, str):
+            raise ValueError()
```